pyro/contrib/named.py

```diff
@@ -49,37 +77,71 @@ def __getattribute__(self, key):
 
     def __setattr__(self, key, value):
         if isinstance(value, (List, Dict)):
-            value._set_name("{}.{}".format(self, key))
+            value._set_name("{}.{}".format(self._name, key))
         if hasattr(self, key):
             old = super(Object, self).__getattribute__(key)
             if not isinstance(old, Object) or not old._is_placeholder:
-                raise RuntimeError("Cannot overwrite {}.{}".format(self, key))
+                raise RuntimeError("Cannot overwrite {}.{}".format(self._name, key))
         super(Object, self).__setattr__(key, value)
 
     @functools.wraps(pyro.sample)
     def sample_(self, fn, *args, **kwargs):
         if not self._is_placeholder:
-            raise RuntimeError("Cannot .sample_ an initialized named.Object {}".format(self))
-        value = pyro.sample(str(self), fn, *args, **kwargs)
+            raise RuntimeError("Cannot .sample_ an initialized named.Object {}".format(self._name))
+        value = pyro.sample(self._name, fn, *args, **kwargs)
         self._set_value(value)
         return value
 
     @functools.wraps(pyro.observe)
     def observe_(self, fn, obs, *args, **kwargs):
         if not self._is_placeholder:
-            raise RuntimeError("Cannot .observe_ an initialized named.Object {}".format(self))
-        value = pyro.observe(str(self), fn, obs, *args, **kwargs)
+            raise RuntimeError("Cannot .observe_ an initialized named.Object {}".format(self._name))
+        value = pyro.observe(self._name, fn, obs, *args, **kwargs)
         self._set_value(value)
         return value
 
     @functools.wraps(pyro.param)
     def param_(self, *args, **kwargs):
         if not self._is_placeholder:
             raise RuntimeError("Cannot .param_ an initialized named.Object")
-        value = pyro.param(str(self), *args, **kwargs)
+        value = pyro.param(self._name, *args, **kwargs)
         self._set_value(value)
         return value
 
+    def set_(self, value):
+        if not self._is_placeholder:
+            raise RuntimeError("Cannot .set_ an initialized named.Object")
+        self._set_value(value)
+        return value
+
+    @contextlib.contextmanager
+    def iarange_(self, *args, **kwargs):
+        if not self._is_placeholder:
+            raise RuntimeError("Cannot .iarange_ an initialized named.Object")
+
+        # Yields both a subsampled data and an indexed latent object.
```

Comment:
Interesting design choice!

Reply:
This should match the behavior of `pyro.iarange`:
```py
@functools.wraps(pyro.iarange)
def iarange_(self, *args, **kwargs):
    return pyro.iarange(self._name, *args, **kwargs)  # Returns a context manager.
```
Feel free to implement a differently-named helper that does what you want, e.g.
```py
def iaenumerate_(self, *args, **kwargs):
    with pyro.iarange(self.iarange._name, *args, **kwargs) as ind:
        yield ind, self.data
```